synapse/federation/federation_server.py

```diff
@@ -135,22 +141,41 @@ async def on_incoming_transaction(
         request_time = self._clock.time_msec()
 
         transaction = Transaction(**transaction_data)
+        transaction_id = transaction.transaction_id  # type: ignore
 
-        if not transaction.transaction_id:  # type: ignore
+        if not transaction_id:
             raise Exception("Transaction missing transaction_id")
 
-        logger.debug("[%s] Got transaction", transaction.transaction_id)  # type: ignore
+        logger.debug("[%s] Got transaction", transaction_id)
 
-        # use a linearizer to ensure that we don't process the same transaction
-        # multiple times in parallel.
+        # We wrap in a ResponseCache so that we de-duplicate retried
+        # transactions.
+        return await self._transaciton_resp_cache.wrap(
+            (origin, transaction_id),
+            self._on_incoming_transaction_inner,
+            origin,
+            transaction,
+            request_time,
+        )
+
+    async def _on_incoming_transaction_inner(
+        self, origin: str, transaction: Transaction, request_time: int
+    ) -> Tuple[int, Dict[str, Any]]:
+        # Use a linearizer to ensure that process transactions from a remote in
+        # order.
         with (
             await self._transaction_linearizer.queue(
                 (origin, transaction.transaction_id)  # type: ignore
```

Comment:
One of the ideas in #8295 was to queue based on origin only. I think we need that change or we'll always have queues of 1-item due to de-duplicating the transaction ID?

Reply:
Woops, that got lost

Reply:
Done